src/Data/Text/Encoding.hs

```diff
@@ -213,45 +218,122 @@ isValidBS bs = start 0
 #endif
 #endif
 
--- | Decode a 'ByteString' containing UTF-8 encoded text.
---
--- Surrogate code points in replacement character returned by 'OnDecodeError'
--- will be automatically remapped to the replacement char @U+FFFD@.
-decodeUtf8With ::
-#if defined(ASSERTS)
-  HasCallStack =>
-#endif
-  OnDecodeError -> ByteString -> Text
-decodeUtf8With onErr bs
-  | isValidBS bs =
-    let !(SBS.SBS arr) = SBS.toShort bs in
-      (Text (A.ByteArray arr) 0 (B.length bs))
-  | B.null undecoded = txt
-  | otherwise = txt `append` (case onErr desc (Just (B.head undecoded)) of
-    Nothing -> txt'
-    Just c  -> T.singleton c `append` txt')
-  where
-    (txt, undecoded) = decodeUtf8With2 onErr mempty bs
-    txt' = decodeUtf8With onErr (B.tail undecoded)
-    desc = "Data.Text.Internal.Encoding: Invalid UTF-8 stream"
-
--- | Decode two consecutive bytestrings, returning Text and undecoded remainder.
-decodeUtf8With2 ::
-#if defined(ASSERTS)
-  HasCallStack =>
-#endif
-  OnDecodeError -> ByteString -> ByteString -> (Text, ByteString)
-decodeUtf8With2 onErr bs1@(B.length -> len1) bs2@(B.length -> len2) = runST $ do
-  marr <- A.new len'
-  outer marr len' 0 0
+data Progression
+  = WriteAndAdvance Char (Int -> Int)
+  | NeedMore
+  | Invalid
+
+decodeChunks :: (Bits w, Num w, Storable w) =>
+  w -- only used for Storable.sizeOf argument which the function discards
+  -> (ByteString -> ByteString -> Int -> Maybe (A.Array, Int))
+  -> ((Int -> Word8) -> Int -> Int -> Progression)
```

Comment:
We can get rid of this continuation. Now that `decodeChunks` returns a `DecodeResult`, the caller can directly do the work of that continuation (including error recovery on invalid bytes), and then call `decodeChunks` again on the remaining chunk.

Reply:
That's not a continuation. That's the function that dictates by what rules to decode the `ByteStream` chunks into `Text`. `decodeChunks` is not exposed outside the module. It's just decoding processes common to all the chunk decoders. In fact, chunk decoders can be for ASCII, Latin-1, and CP-1252 just with providing the rules via the `decodeF` (and optionally `queryOptimization`).

Reply:
Sorry I had a brainfart moment here!